drools-parser/src/main/java/org/drools/parser/DRLVisitorImpl.java

```diff
@@ -342,31 +349,8 @@ private List<BaseDescr> visitDescrChildren(RuleNode node) {
     }
 
     // leaves of constraint concatenate return Strings
-    private String visitConstraintChildren(RuleNode node) {
-        return ((ParserRuleContext) node).children.stream()
-                .map(c -> c instanceof TerminalNode ? c : c.accept(this))
-                .filter(Objects::nonNull)
-                .map(Object::toString)
-                .collect(Collectors.joining(" "));
-    }
-
-    @Override
-    public Object visitChildren(RuleNode node) {
-        if (hasConstraintAsAncestor(node)) {
-            return visitConstraintChildren(node);
-        }
-        return super.visitChildren(node);
-    }
-
-    private boolean hasConstraintAsAncestor(RuleNode node) {
-        ParseTree parent = node.getParent();
-        if (parent instanceof DRLParser.ConstraintContext) {
-            return true;
-        } else if (parent == null) {
-            return false;
-        } else {
-            return hasConstraintAsAncestor((RuleNode) parent);
-        }
+    private String visitConstraintChildren(ParserRuleContext ctx) {
+        return getTokenTextPreservingWhitespace(ctx, tokenStream);
```

Comment:
Just get a constraint string via tokenStream rather than collecting tokens with a delimiter ` ` (it breaks a test). This is the same approach as the current Drools parser implementation.